Use GIL to guard decref of jit::toPyObj return value in processRpc

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38415 Enforce const on PyRRef functions
* #38402 Remove unnecessary py::object copy in PyRRef ctor
* **#38376 Use GIL to guard decref of jit::toPyObj return value in processRpc**
* #38348 Use GIL to guard py::object's destruction
* #38366 Explicitly decref py::object in PythonRpcHandler
* #38364 Explicitly decref py::object in ConcretePyObjectHolder and PythonFunctionGuard
* #38340 Minor code cleanup



Differential Revision: [D21540179](https://our.internmc.facebook.com/intern/diff/D21540179)

Commit:
Use GIL to guard decref of jit::toPyObj return value in processRpc

[ghstack-poisoned]

Commit:
Update on "Use GIL to guard decref of jit::toPyObj return value in processRpc"


Differential Revision: [D21540179](https://our.internmc.facebook.com/intern/diff/D21540179)

[ghstack-poisoned]

xush6528 commented:
There are more sites, right? MessageType::PYTHON_CALL, and MessageType::PYTHON_REMOTE_CALL.

xush6528 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -33,9 +33,11 @@ using namespace torch::distributed::autograd;
 namespace {
 
-py::object toPyObj(IValue value) {
+SerializedPyObj toSerializedPyObj(IValue value) {
```
It's only used by `MessageType::PYTHON_RREF_FETCH_CALL`.

Can we move under case `MessageType::PYTHON_RREF_FETCH_CALL` and make it a lambda, just like the markComplete lambda on line 116.

mrshenli replied:
Sure, edited, but the downside is that we will need to capture this lambda in two nested callbacks. I am not sure if the extra overhead worth it here.

xush6528 replied:
@mrshenli 

> If there is no capture by value, the closure type is typically trivially copyable and trivially moveable.
http://www.cplusplus.com/forum/general/202662/

Commit:
Update on "Use GIL to guard decref of jit::toPyObj return value in processRpc"


Differential Revision: [D21540179](https://our.internmc.facebook.com/intern/diff/D21540179)

[ghstack-poisoned]

Commit:
Update on "Use GIL to guard decref of jit::toPyObj return value in processRpc"




Differential Revision: [D21540179](https://our.internmc.facebook.com/intern/diff/D21540179)

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -354,6 +349,13 @@ void RequestCallbackImpl::processRpc(
+        auto& pythonRpcHandler = PythonRpcHandler::getInstance();
+        // Need this GIL to guard jit::toPyObj and destruct its returned
+        // py::object
+        pybind11::gil_scoped_acquire ag;
```
nit: `pybind11::gil_scoped_acquire ag;` -> `py::gil_scoped_acquire ag;`

Commit:
Update on "Use GIL to guard decref of jit::toPyObj return value in processRpc"




Differential Revision: [D21540179](https://our.internmc.facebook.com/intern/diff/D21540179)

[ghstack-poisoned]

Commit:
Update on "Use GIL to guard decref of jit::toPyObj return value in processRpc"




Differential Revision: [D21540179](https://our.internmc.facebook.com/intern/diff/D21540179)

[ghstack-poisoned]

Commit:
Update on "Use GIL to guard decref of jit::toPyObj return value in processRpc"




Differential Revision: [D21540179](https://our.internmc.facebook.com/intern/diff/D21540179)

[ghstack-poisoned]
